### ISSUE TITLE
Patreon scraping issues

### ISSUE THREAD

**Issue by 4EC2A0** [fake-name/xA-Scraper#14]
I'm having a couple issues with trying to scrape from Patreon.

The first is that I'm getting a bunch of errors talking about an invalid dlstate `not_set` They're all pretty much the same, and look like this

```
ERROR - Unknown error in page retrieval!
ERROR - PostID = '[PostID]'
ERROR - Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
psycopg2.DataError: invalid input value for enum dlstate_enum: "not_set"
LINE 1: ...eta, state, filename, fspath) VALUES ([ID], 0, '', 'not_set',...
                                                             ^


The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/user/xA-Scraper-master/rewrite/modules/patreon/patreonScrape.py", line 335, in getArtist
    postTags=ret['post_tags'],
  File "/home/user/xA-Scraper-master/rewrite/modules/scraper_base.py", line 212, in _updatePreviouslyRetreived
    sess.commit()
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 906, in commit
    self.transaction.commit()
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 461, in commit
    self._prepare_impl()
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 441, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 2177, in flush
    self._flush(objects)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 2297, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/util/compat.py", line 187, in reraise
    raise value
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/session.py", line 2261, in _flush
    flush_context.execute()
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/unitofwork.py", line 389, in execute
    rec.execute(self)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/unitofwork.py", line 548, in execute
    uow
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/persistence.py", line 181, in save_obj
    mapper, table, insert)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/orm/persistence.py", line 835, in _emit_insert_statements
    execute(statement, params)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/util/compat.py", line 186, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/usr/local/lib/python3.5/dist-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DataError: (psycopg2.DataError) invalid input value for enum dlstate_enum: "not_set"
LINE 1: ...eta, state, filename, fspath) VALUES ([ID], 0, '', 'not_set',...
                                                             ^
 [SQL: 'INSERT INTO art_file (item_id, seqnum, file_meta, state, filename, fspath) VALUES (%(item_id)s, %(seqnum)s, %(file_meta)s, %(state)s, %(filename)s, %(fspath)s) RETURNING art_file.id'] [parameters: {'filename': None, 'state': 'not_set', 'seqnum': 0, 'file_meta': '', 'item_id': [ID], 'fspath': '/home/user/xAS-Downloads/Patreon/[Creator]/[Filename]'}]
```

At the end when (I assume) it's trying to insert everything into the database it fails saying `CRITICAL - Uncaught exception` and `sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (psycopg2.DataError) invalid input value for enum dlstate_enum: "not_set"` and nothing seems to get added.

The other issue is that it doesn't seem to find images on posts that don't have them as attachments. I'm not sure what the specific differences are, but for these posts the image isn't listed down at the bottom as an attachment, but it's still there at the top.

Edit: Also, just to clarify for the attachment issue, I did make sure that those images aren't actually being downloaded. Ex. one of the creators I follow that has ~700 posts (not all, but mostly images), and the download folder only has 4 files.

**Comment by fake-name** [comment]
Ok, so the problem appears to be the giant rewrite I did recently. You're running the old database schema, with the new code. 

It *should* be possible to update the schema without any data loss, but there have been strange issues in the migration, see https://github.com/fake-name/xA-Scraper/issues/13.

TL;DR try running `python3 db_migrate.py db upgrade --tag force-migrate` from the repository root to force the db upgrade. PLEASE back up the database first, @mmsterful has reported that the migration has accidentally eaten his DB content due to dumb bugs on my part in the past.


------

I'm not sure what's going on about non-attachment images. It certainly seems to fetch them for me, though it could have broken recently. I will check.

**Comment by 4EC2A0** [comment]
This is my first time setting this up, maybe I did something wrong? I don't have any data yet so I'm not worried about that. Here's what I did:

- Install postgresql, set up user/pass and allow on localhost. Created the database with `create database xas;`
- Install all the python libs.
- Download/unzip the repo
- Modify settings.base.py and copy it to settings.py. The only things I changed were the postgres settings, the Patreon settings, and the download/content paths. I wasn't sure if I was supposed/needed to do anything with the rpc or captcha settings.
- run `./run.sh`
- I can connect to the web interface/log in. Although things that I noticed were
    - Error Lists page says`Invalid site-name! The site-name 'errLists' is not in the valid site-name list ['da', 'fa', 'hf', 'wy', 'ib', 'px', 'sf', 'pat', 'yp']`
    - all of the individual lists say `File Not Found`
    - and the upload settings page also says `File Not Found`
- I tried with and without running `python3 db_migrate.py db upgrade --tag force-migrate` before trying to fetch with `python3 -m manage fetch pat`

I'm still seeing the same problems.

**Comment by fake-name** [comment]
Oh, this is a new setup? Huh.

I recently did a big rewrite, moving from some really, REALLY old and horrible code I wrote when I was just learning python to something at least marginally sane, so a fair bit of the codebase is still somewhat broken or not-implemented  (the scrapers are functional, a lot of the auxiliary stuff/web-UI isn't).

Anyways, if you just want the patreon scraping facilities, you might try rolling back to https://github.com/fake-name/xA-Scraper/commit/2aba2f7b4f27fdd3008d5eee7656fc2762808eeb. That's pre-migration, but it was functional. 

------

Anyways, I think if you delete the contents of the database, and re-run `./run.sh`, it should now work. My suspicion is that the migration wasn't capturing alterations to one of the database enums, which I just corrected retrospectively in 4482c9cc0d6831427af10f8a2b35847f21262939.

**Comment by fake-name** [comment]
Ok, so yeah, patreon moved where they listed non-attachment post files. It should now be fixed again. Thanks for letting me know about that! I had missed it myself.

**Comment by 4EC2A0** [comment]
Haha, I was right about to say that it looks like they changed API. The other fix for the database worked too. Everything seems to be working fine now. Thanks :)

Oh, also another completely unrelated problem I had before. I was getting a missing library error before about `mprpc`, I had to add it to requirements.txt to get it running.

**Comment by fake-name** [comment]
Awesome! 

The `mprpc` thing is more recent changes I haven't had a chance to propagate down to the install procedure. Sorry about that.
